### ISSUE TITLE
superclass mismatch for class ThreadSafe

### ISSUE THREAD

**Issue by cappadona** [locomotivecms/wagon#255]
Attempting to setup local development of wagon and steam and first ensuring I can test server with the default site (spec/features/default). I'm using MRI 2.2.2.

### Scenario 1: Passing Test

* local wagon-2.0.0-beta.2 (locomotivecms/wagon@2b9285c)
* steam-1.0.0-beta3 via [rubygems](https://rubygems.org/gems/locomotivecms_steam/versions/1.0.0.pre.beta.3)

###### Steps to reproduce
1. Clone wagon repo
1. `$ cd <wagon-clone> && bundle install`
1. `$ bundle exec bin/wagon serve -v spec/fixtures/default`

```ruby
Thin web server (v1.6.3 codename Protein Powder)
Maximum connections set to 1024
Listening on 0.0.0.0:3333, CTRL+C to stop
```

### Scenario 2: Failing Test

* local wagon-2.0.0-beta.2 (locomotivecms/wagon@2b9285c)
* local steam-1.0.0-beta-3 (locomotivecms/steam@2998aff)

###### Steps to reproduce
1. Clone each repo
1. `$ cd <steam-clone>`
1. Checkout 1.0.0-beta-3 (as of right now there have been [9 commits to master since 1.0.0-beta-3](https://github.com/locomotivecms/steam/compare/v1.0.0.pre.beta.3...f5c23e0))
 
  ```shellsession
$ git checkout v1.0.0.pre.beta.3
```
> **NOTE:** keeping local steam at latest (locomotivecms/steam@f5c23e0) has no effect
1. `$ cd ../<wagon-clone>`
1. Update Gemfile to point to local steam, here's the diff:
   ```diff
diff --git a/Gemfile b/Gemfile
index 85774f6..c9d1c84 100644
--- a/Gemfile
+++ b/Gemfile
@@ -14,7 +14,7 @@ gem 'therubyracer'
 
 # Local development
 # gem 'locomotivecms_coal', path: '../in_progress/coal', require: false
-# gem 'locomotivecms_steam', path: '../in_progress/steam', require: false
+gem 'locomotivecms_steam', path: '/usr/local/src/locomotivecms_steam', require: false
 # gem 'locomotivecms_common', path: '../in_progress/common', require: false
 
 group :development, :test do
```
1. `$ bundle install`
1. `$ bundle exec bin/wagon serve -v spec/fixtures/default`

   ```ruby
superclass mismatch for class ThreadSafe
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/threadsafe.rb:3:in `<module:Middlewares>'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/threadsafe.rb:1:in `<top (required)>'
  /usr/local/src/locomotivecms_wagon/lib/locomotive/wagon/middlewares/error_page.rb:2:in `require'
  /usr/local/src/locomotivecms_wagon/lib/locomotive/wagon/middlewares/error_page.rb:2:in `<top (required)>'
  /usr/local/src/locomotivecms_wagon/lib/locomotive/wagon/commands/serve_command.rb:51:in `require'
  /usr/local/src/locomotivecms_wagon/lib/locomotive/wagon/commands/serve_command.rb:51:in `require_steam'
  /usr/local/src/locomotivecms_wagon/lib/locomotive/wagon/commands/serve_command.rb:26:in `start'
  /usr/local/src/locomotivecms_wagon/lib/locomotive/wagon/commands/serve_command.rb:14:in `start'
  /usr/local/src/locomotivecms_wagon/lib/locomotive/wagon.rb:39:in `serve'
  /usr/local/src/locomotivecms_wagon/lib/locomotive/wagon/cli.rb:277:in `serve'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
  bin/wagon:12:in `<main>'
```

**Comment by cappadona** [comment]
Not sure if this is the same underlying issue as #254, but picking up from **scenario 2** above, commenting out all references to `ThreadSafe` in wagon allows the Thin web server to start without error, but the first request to `http://localhost:3333` results in:

   ```ruby
Unexpected error while processing request: superclass mismatch for class MemoryAdapter
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/adapters/memory.rb:9:in `<module:Steam>'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/adapters/memory.rb:7:in `<top (required)>'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/adapters/filesystem.rb:1:in `require_relative'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/adapters/filesystem.rb:1:in `<top (required)>'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/repositories.rb:44:in `require_relative'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/repositories.rb:44:in `build_adapter'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/repositories.rb:11:in `block in <class:Repositories>'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/morphine-0.1.1/lib/morphine.rb:15:in `instance_exec'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/morphine-0.1.1/lib/morphine.rb:15:in `block in register'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/repositories.rb:15:in `block in <class:Repositories>'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/morphine-0.1.1/lib/morphine.rb:15:in `instance_exec'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/morphine-0.1.1/lib/morphine.rb:15:in `block in register'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/services.rb:44:in `block in <class:Instance>'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/morphine-0.1.1/lib/morphine.rb:15:in `instance_exec'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/morphine-0.1.1/lib/morphine.rb:15:in `block in register'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/services.rb:36:in `block (2 levels) in <class:Instance>'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/services.rb:24:in `call'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/services.rb:24:in `__getobj__'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/services.rb:27:in `nil?'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/site.rb:14:in `_call'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/threadsafe.rb:12:in `call'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/default_env.rb:13:in `call'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:225:in `context'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:220:in `call'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/lint.rb:49:in `_call'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/lint.rb:37:in `call'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/dynamic_assets.rb:20:in `call'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/static.rb:124:in `call'
  /usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/favicon.rb:13:in `call'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-rewrite-1.5.1/lib/rack/rewrite.rb:24:in `call'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/builder.rb:153:in `call'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thin-1.6.3/lib/thin/connection.rb:86:in `block in pre_process'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thin-1.6.3/lib/thin/connection.rb:84:in `catch'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thin-1.6.3/lib/thin/connection.rb:84:in `pre_process'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/eventmachine-1.0.8/lib/eventmachine.rb:1062:in `call'
  ~/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/eventmachine-1.0.8/lib/eventmachine.rb:1062:in `block in spawn_threadpool'
```


###### Diff

   ```diff
diff --git a/lib/locomotive/wagon/commands/serve_command.rb b/lib/locomotive/wagon/commands/serve_command.rb
index 5576612..67aee71 100644
--- a/lib/locomotive/wagon/commands/serve_command.rb
+++ b/lib/locomotive/wagon/commands/serve_command.rb
@@ -64,7 +64,7 @@ module Locomotive::Wagon
           config.middleware.insert_before Rack::Lint, Rack::LiveReload, live_reload_port: port
         end
 
-        config.middleware.insert_before Rack::Lint, Locomotive::Wagon::Middlewares::ErrorPage
+        # config.middleware.insert_before Rack::Lint, Locomotive::Wagon::Middlewares::ErrorPage
       end
     end
```

   ```diff
diff --git a/lib/locomotive/wagon/middlewares/error_page.rb b/lib/locomotive/wagon/middlewares/error_page.rb
index 00ce140..a47fffe 100644
--- a/lib/locomotive/wagon/middlewares/error_page.rb
+++ b/lib/locomotive/wagon/middlewares/error_page.rb
@@ -1,72 +1,72 @@
-require 'erb'
-require 'locomotive/steam/middlewares/threadsafe'
-require 'locomotive/steam/middlewares/helpers'
-
-module Locomotive::Wagon
-  module Middlewares
-
-    # Display a nice page error
-    #
-    class ErrorPage < Locomotive::Steam::Middlewares::ThreadSafe
-
-      include Locomotive::Steam::Middlewares::Helpers
-
-      def _call
-        begin
-          self.next
-        rescue StandardError => error
-          @error = error
-          log_error
-          render_error_page
-        end
-      end
-
-      private
-
-      def log_error
-        log "Error: #{@error.message}".red
-        log @error.backtrace.join("\n")
-      end
-
-      def render_error_page
-        _template = ERB.new(template, nil, '-')
-        render_response(_template.result(binding))
-      end
-
-      def template
-        %{
-<DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
-
-  <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
-    <head>
-      <title>Wagon - Rendering error</title>
-    </head>
-
-    <body>
-      <h1>Arrrghhhh, we could not render page</h1>
-      <h2><%= @error.message %></h2>
-
-      <h3>File: <%= @error.respond_to?(:file) ? @error.file : '?' %></h3>
-
-      <h3>Code</h3>
-      <% if @error.respond_to?(:code_lines) %>
-        <pre>
-<% @error.code_lines.each do |(line, statement)| -%>
-<strong><%= line %></strong> <%= statement %>
-<% end -%>
-        </pre>
-      <% else %>
-        <p><i>No code</i></p>
-      <% end %>
-
-      <h3>Back trace</h3>
-      <%= @error.backtrace.join("<br/>") %>
-
-    </body>
-  </html>}
-      end
-
-    end
-
-  end
-end
+# require 'erb'
+# require 'locomotive/steam/middlewares/threadsafe'
+# require 'locomotive/steam/middlewares/helpers'
+
+# module Locomotive::Wagon
+#   module Middlewares
+
+#     # Display a nice page error
+#     #
+#     class ErrorPage < Locomotive::Steam::Middlewares::ThreadSafe
+
+#       include Locomotive::Steam::Middlewares::Helpers
+
+#       def _call
+#         begin
+#           self.next
+#         rescue StandardError => error
+#           @error = error
+#           log_error
+#           render_error_page
+#         end
+#       end
+
+#       private
+
+#       def log_error
+#         log "Error: #{@error.message}".red
+#         log @error.backtrace.join("\n")
+#       end
+
+#       def render_error_page
+#         _template = ERB.new(template, nil, '-')
+#         render_response(_template.result(binding))
+#       end
+
+#       def template
+#         %{
+# <DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+
+#   <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
+#     <head>
+#       <title>Wagon - Rendering error</title>
+#     </head>
+
+#     <body>
+#       <h1>Arrrghhhh, we could not render page</h1>
+#       <h2><%= @error.message %></h2>
+
+#       <h3>File: <%= @error.respond_to?(:file) ? @error.file : '?' %></h3>
+
+#       <h3>Code</h3>
+#       <% if @error.respond_to?(:code_lines) %>
+#         <pre>
+# <% @error.code_lines.each do |(line, statement)| -%>
+# <strong><%= line %></strong> <%= statement %>
+# <% end -%>
+#         </pre>
+#       <% else %>
+#         <p><i>No code</i></p>
+#       <% end %>
+
+#       <h3>Back trace</h3>
+#       <%= @error.backtrace.join("<br/>") %>
+
+#     </body>
+#   </html>}
+#       end
+
+#     end
+
+#   end
+# end
```

**Comment by did** [comment]
hi @cappadona, thanks for the very precise description of the bug. Absolutely no idea what's going on here.

**Comment by did** [comment]
**My environment**
- local version of Wagon (c8d0cd2c1443f93f6b5972456d944109c4a2fd42)
- local version of Steam (f5c23e0fc77eb052ea780154dafc508db461048f)
- ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-darwin14]

**Command**
```
bundle exec bin/wagon serve spec/fixtures/default
````
~/Documents/LocomotiveCMS/wagon[master] > be bin/wagon serve spec/fixtures/default
Thin web server (v1.6.3 codename Protein Powder)
Maximum connections set to 1024
Listening on 0.0.0.0:3333, CTRL+C to stop

**No error**

**Comment by cappadona** [comment]
Thanks for the prompt response, @did. Any chance you have some local commits that you haven't pushed to GitHub?

I just searched the entire log for Wagon and can't find any commit hash that begins with c8d0cd2.

**Comment by did** [comment]
@cappadona I'm pushing my last commit (nothing fancy)

**Comment by did** [comment]
@cappadona done

**Comment by cappadona** [comment]
Thanks @did. No change in behavior on my end, and not a surprise.

No error when using steam from rubygems or GitHub, but if I point Wagon's Gemfile to my local steam clone I get the `superclass mismatch`. Strange. I'm also using `ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-darwin14]`

I'll let you know if I get lucky...

**Comment by did** [comment]
interesting...
http://stackoverflow.com/questions/9785694/superclass-mismatch-struct-reloading-and-spork

The question, in our case, is why the hell Wagon is reloading the class in your version of Wagon/Steam and not in mine.

**Comment by cappadona** [comment]
Nice find...yeah, and why is Wagon only reloading the class when I reference the local Steam?

**Comment by did** [comment]
Again no idea at all! btw, could you put "puts" in the MemoryAdapter and see it's called twice?
What I noticed over the years is that ruby files are not loaded in the same order depending on the machine / *nix. Is that related to your problem? Hmmm, maybe.
Thanks for your help.

**Comment by cappadona** [comment]
`puts` in MemoryAdapter is only called once.

Making the following changes as suggested in [one of the answers](http://stackoverflow.com/a/9787113) on the SO thread you referenced above (as well as [here](http://www.benjaminoakes.com/2013/10/15/superclass-mismatch-when-inheriting-from-struct-new)) allows me to serve the default site and actually successfully respond to requests:

   ```diff
diff --git a/lib/locomotive/steam/adapters/memory.rb b/lib/locomotive/steam/adapters/memory.rb
index d321d8a..5110b23 100644
--- a/lib/locomotive/steam/adapters/memory.rb
+++ b/lib/locomotive/steam/adapters/memory.rb
@@ -6,7 +6,7 @@ require_relative 'memory/dataset'
 
 module Locomotive::Steam
 
-  class MemoryAdapter < Struct.new(:collection)
+  MemoryAdapter = Struct.new(:collection) do
 
     include Locomotive::Steam::Adapters::Concerns::Key
```

   ```diff
diff --git a/lib/locomotive/steam/middlewares/threadsafe.rb b/lib/locomotive/steam/middlewares/threadsafe.rb
index a29ad37..4fa2586 100644
--- a/lib/locomotive/steam/middlewares/threadsafe.rb
+++ b/lib/locomotive/steam/middlewares/threadsafe.rb
@@ -1,6 +1,6 @@
 module Locomotive::Steam::Middlewares
 
-  class ThreadSafe < Struct.new(:app)
+  ThreadSafe = Struct.new(:app) do
 
     attr_accessor :env
```

**Comment by did** [comment]
Awesome! Let's get rid of these useless ````Struct.new````! (http://thepugautomatic.com/2013/08/struct-inheritance-is-overused/).
thanks!

**Comment by cappadona** [comment]
Cool...just FYI I still do get some warnings:

   ```ruby
/usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/threadsafe.rb:3: warning: already initialized constant Locomotive::Steam::Middlewares::ThreadSafe
/usr/local/src/locomotivecms_steam/lib/locomotive/steam/middlewares/threadsafe.rb:3: warning: previous definition of ThreadSafe was here
Thin web server (v1.6.3 codename Protein Powder)
Maximum connections set to 1024
Listening on 0.0.0.0:3333, CTRL+C to stop
/usr/local/src/locomotivecms_steam/lib/locomotive/steam/adapters/memory.rb:9: warning: already initialized constant Locomotive::Steam::MemoryAdapter
/usr/local/src/locomotivecms_steam/lib/locomotive/steam/adapters/memory.rb:9: warning: previous definition of MemoryAdapter was here
```

**Comment by did** [comment]
That’s exactly what I was going to ask you!
My fix might get rid of these warnings but not sure about that. 

Hmmm, wait, I’m thinking about something here. The name of these 2 files doesn’t reflect the class itself. I mean, threadsafe.rb should be renamed into “thread_safe.rb”. memory.rb does not reference a Memory module inside it. Alright, that’s the bug! 

Let me fix that right now.

**Comment by did** [comment]
@cappadona could you test with this commit please (Steam) locomotivecms/steam@https://github.com/locomotivecms/steam/commit/a7b7c6fbb6eb946d955bfdba051ab1b6981bf9e1. thanks!

**Comment by cappadona** [comment]
Not quite...I needed to update Wagon as well otherwise I get a `cannot load such file -- locomotive/steam/middlewares/threadsafe`:

   ```diff
diff --git a/lib/locomotive/wagon/middlewares/error_page.rb b/lib/locomotive/wagon/middlewares/error_page.rb
index 00ce140..0673e50 100644
--- a/lib/locomotive/wagon/middlewares/error_page.rb
+++ b/lib/locomotive/wagon/middlewares/error_page.rb
@@ -1,5 +1,5 @@
 require 'erb'
-require 'locomotive/steam/middlewares/threadsafe'
+require 'locomotive/steam/middlewares/thread_safe'
 require 'locomotive/steam/middlewares/helpers'
 
 module Locomotive::Wagon
```

Unfortunately, after making this change, I'm right back where I started. If my Wagon Gemfile points to rubygems or GitHub, no errors, but change the reference to the same commit on my local clone and I get the `superclass mismatch` once again.

**Comment by did** [comment]
what if you apply your quick fix as described above on my latest commit? Will that remove the warnings? Sorry to bother you but since I can't reproduce it, you're my only hope!

**Comment by cappadona** [comment]
No bother at all. I appreciate your help.

Sorry for the delay...just submitted two PRs.

**Comment by cappadona** [comment]
Hey @did, you're not going to believe this (or maybe you will?) but I figured out the reason why you couldn't reproduce this behavior locally. Turns out [symlinks can trigger double loads for require](https://bugs.ruby-lang.org/issues/4403).

I have `/usr/local/src` symlinked to my Dropbox and seeing the change in path in the warnings from `wagon serve -v` after our latest changes, piqued my interest. I confirmed that the symlink is the culprit by checking out wagon-2.0.0-beta.2 (locomotivecms/wagon@2b9285c) and steam-1.0.0-beta-3 (locomotivecms/steam@2998aff) in my local clones and using either a relative or absolute path for Steam in Wagon's Gemfile. When avoiding the symlink, I was able to reproduce your behavior (**no error**).

I suspect that this may be the underlying issue for @KazW as well in #254. 

I found a couple suggestions to dealing with this, such as [Pathname.realpath](http://ruby-doc.org/stdlib-2.2.2/libdoc/pathname/rdoc/Pathname.html#method-i-realpath):
http://stackoverflow.com/questions/322769/finding-relative-libraries-when-using-symlinks-to-ruby-executables

But perhaps the best approach is to just avoid symlinks when pointing to local gems. Sorry for the hassle and thanks for your help.
